contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,439 @@
+allow for altering subsriptions, which should be very simple once we have list
+implemented. But since websocket API is needed anyway, we also dropped it.*
+
+
+### Filtering definition
+
+ We need to make our filtering mechanism to support different resource types at
+the same time. On the other hand, we would like it to be as consistent with all
+other Kubernetes APIs as possible. So we define the selector for bulk operations
+as following:
+
+```
+type GetOperations struct {
+	Operations []GetOperation
+}
+
+type GetOperation struct {
+	Resource GroupVersionResource
+
+	// We would like to reuse the same ListOptions definition as we are using
+	// in regular APIs.
+	// TODO: We may consider supporting multiple ListOptions for a single
+	// GetOperation. 
+	Options ListOptions
+}
+```
+
+ We need to be able to detect whether a given user is allowed to get/list
+objects requested by a given "GetOperation". For that purpose, we will create
+some dedicated admission plugin (or potentially reuse already existing one).
+That one will not support partial rejections and will simply allow or reject
+the whole request.
+
+For watch operations, as described in requirements, we also need to periodically
+(or maybe lazily) verify whether their ACLs didn't change (and user is still
+allowed to watch requested objects). However, as also mentioned above, this
+periodic checking isn't specific to bulk operations and we need to support the
+same mechanism for regular watch too. We will just ensure that this mechanism
+tracking and periodically verifying ACLs is implemented low enough in apiserver
+machinery so that we will be able to reuse exactly the same one for the purpose
+of bulk watch operations.
+For watch request, we will support partial rejections. The exact details of it
+will be described together with dynamic watch description below.
+
+
+### Dynamic watch
+
+ As mentioned in the Proposal section, we will implement bulk watch that will
+allow for dynamic subscribtion/unsubscribtion for (sets of) objects on top of
+websockets protocol.
+
+ Note that we already support websockets in the regular Kubernetes API for
+watch requests (in addition to regular http requests), so for the purpose of
+bulk watch we will be extending websocket support.
+
+ The the high level, the propocol will look:
+1. client opens a new websocket connection to a bulk watch endpoint to the
+server via ghttp GET
+1. this results in creating a single channel that is used only to handle
+communication for subscribing/unsubscribing for watches; no watch events are
+delivered via this particular channel.  
+*TODO: Consider switching to two channels, one for incoming and one for
+outgoing communication*
+1. to subscribe for a watch of a given (set of) objects, user sends `Watch`
+object over the channel; in response a new channel is created and the message
+with the channel identifier is send back to the user (we will be using integers
+as channel identifiers).
```

Comment:
TODO: check if integers are mandatory or if we could do something like "ch1", "ch2"...